tidy3d/components/lumped_element.py

```diff
@@ -3,100 +3,151 @@
+    material properties given their size, voltage axis, and the network they represent."""
 
-    voltage_axis: Axis = pydantic.Field(
+    voltage_axis: Axis = pd.Field(
         ...,
         title="Voltage Drop Axis",
         description="Specifies the axis along which the component is oriented and along which the "
         "associated voltage drop will occur. Must be in the plane of the element.",
     )
 
-    _plane_validator = assert_plane()
+    snap_perimeter_to_grid: bool = pd.Field(
+        True,
+        title="Snap Perimeter to Grid",
+        description="When enabled, the perimeter of the lumped element is snapped to the simulation grid, "
+        "which improves accuracy when the number of grid cells is low within the element. Sides of the element "
+        "perpendicular to the ``voltage_axis`` are snapped to grid boundaries, while the sides parallel to the "
+        "``voltage_axis`` are snapped to grid centers. Lumped elements are always snapped to the nearest grid "
+        "boundary along their ``normal_axis``, regardless of this option.",
+    )
 
     @cached_property
     def normal_axis(self):
-        """Normal axis of the lumped element."""
+        """Normal axis of the lumped element, which is the axis where the element has zero size."""
         return self.size.index(0.0)
 
+    @cached_property
+    def lateral_axis(self):
+        """Lateral axis of the lumped element."""
+        return 3 - self.voltage_axis - self.normal_axis
+
+    @cached_property
+    def _voltage_axis_2d(self) -> Axis2D:
+        """Returns the voltage axis using the in-plane dimensions used by :class:`.Medium2D`."""
+        if self.normal_axis > self.voltage_axis:
+            return self.voltage_axis
+        else:
+            return self.voltage_axis - 1
+
+    @cached_property
+    def _snapping_spec(self) -> SnappingSpec:
+        """Returns the snapping behavior for each dimension of the lumped element.
+
+        Note
+        ----
+
+        Snapping the lumped element is needed for accuracy, since in many cases staircasing
+        will be used instead of subpixel averaging, e.g., when there are many different media
+        in close proximity to the lumped element. The equivalent media produced by the
+        lumped element is usually quite large in magnitude and depends directly on the physical
+        dimensions of the lumped element. As a result, we need to ensure that staircasing will
+        lead to an accurate representation of the lumped element. We make sure to create a
+        :class:`.Box` aligned with the Yee grid that fully encloses the
+        electric field component which is parallel to the ``voltage_axis``.
+        """
+
+        snap_location = [SnapLocation.Boundary] * 3
+        snap_behavior = [SnapBehavior.Closest] * 3
+        snap_location[self.lateral_axis] = SnapLocation.Center
+        snap_behavior[self.lateral_axis] = SnapBehavior.Expand
```

Comment:
I understand that snapping to the center (H-field locations) eliminates staircasing errors in the structure, but why don’t we snap to the center along the normal direction as well? If we use one Yee cell to account for the box thickness, it should be centered around the E-fields. Is my understanding correct?

Reply:
My understanding is that box thickness is 0 along normal direction. It's snapped to the boundary so that Medium2D is not visible to normal E-field.

Reply:
Yeah, that's what I meant— I might have misunderstood the coordinates before. So, the coors along the x-direction are recording the positions for Ex, and it's the same for Ey and Ez, right?

Reply:
right, center coord along x-direction is for Ex. Medium2D is expected to be invisible to the E-field along normal axis, so it's placed on the boundary rather than the center.